core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala

```diff
@@ -1089,6 +1089,85 @@ class AbstractFetcherThreadTest {
     fetcher.verifyLastFetchedEpoch(partition, Some(5))
   }
 
+  @Test
+  def testTruncateOnFetchDoesNotUpdateHighWatermark(): Unit = {
```

Comment:
nit: wonder if we should change the name. The test is not really doing anything with the high watermark; it is just verifying that `processPartitionData` is not called. Maybe a test like this could exist for `ReplicaFetcherThread`?